### ISSUE TITLE
[mlflow] Postgresql as an optional dependency

### ISSUE THREAD

**Issue by ryshoooo** [community-charts/helm-charts#20]
### Is your feature request related to a problem ?

When I deploy the mlflow chart into the cluster, I don't have any postgres database running externally. So I have to deploy the postgres helm chart first and then set the pointers to the database manually. It'd be preferable if the mlflow chart had postgresql dependency available.

### Describe the solution you'd like.

Postgresql helm chart as an optional dependency.

### Describe alternatives you've considered.

Installing postgresql separately or using a cloud-provided database. None of which I'm very happy about.

**Comment by burakince** [comment]
Hi @ryshoooo 

I totally agree with you. I think it can save some time for developers.

On the other hand, there are concerns about running workloads in a persistent state if they were not originally designed as distributed systems. So, running Cockroach DB or Cassandra which are designed to be distributed would be fine, but running more traditional solutions that are centralised with fail-over, which includes Postgres, could be a problem.

Let me try to explain the problem with an example. Let's say we use AWS cloud and EKS in it. And EKS was installed in the `eu-central-1` region. The region doesn't mean a single datacenter. It mostly refers to a city. In our case, it means Frankfurt, Germany. You can find more information [here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html). And availability zones means a datacenter. In our case they are `eu-central-1a`, `eu-central-1b`, and `eu-central-1c`. They are buildings that are physically separated from each other but connected by fast networks. When you try to have a persistent volume in Kubernetes, probably you will use an SSD or HDD storage in the background. It's a physical disc and it will be located in zone `a`, or `b`, or `c`. `Don't put all your eggs in one basket. Because if drop the basket, you will have lost them all at once.` This refers to failover cases in the datacenter. Because of this reason, we use several datacenters for our Kubernetes cluster installations for production usage. In our case, EKS could be created from several nodes which are distributed in `eu-central-1a`, and `eu-central-1b`, and `eu-central-1c` datacenters. But we can lose a single node anytime and maybe our PostgreSQL DB was in that node. So, K8s can decide to run it in different nodes and if you didn't use some special configurations ([K8s Taints and Tolerations](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/)) it can try to run it in a different datacenter. But you can't use the disc from a different datacenter. And if you use taints and tolerations from Kubernetes, then you will always use the same datacenter. And the egg-basket problem will accrue again. Of course, they are some other solutions that exist. e.g. you can use a cloud-native distributed block storage system like [Longhorn](https://longhorn.io/). But it will come with a cost.

So, long story short. I don't recommend PostgreSQL DB on the Kubernetes cluster for production usage. But this dependency could be very useful for personal or test cluster usages. Let's think about design of the values.yaml file.

**Comment by burakince** [comment]
Well, I think we need to support with embedded Postgres and without embedded Postgres (It could be Postgres as a service or VM. Or it could be a different type of DB (e.g. MySQL) in the future.) versions altogether.

The following settings are our current Postgres YAML structure.

```yaml
backendStore:
  postgres:
    enabled: true
    host: "postgresql-instance1.cg034hpkmmjt.eu-central-1.rds.amazonaws.com"
    port: 5432
    database: "mlflow"
    user: "mlflowuser"
    password: "Pa33w0rd!"
```

Maybe we can change it to something like this for backward compatibility. 

```yaml
backendStore:
  embeddedPostgres:
    enabled: true
    ***all settings coming from [bitnami postgresql](https://artifacthub.io/packages/helm/bitnami/postgresql) chart.***
  postgres:
    enabled: false
    host: ""
    port: 5432
    database: ""
    user: ""
    password: ""
```

What do you think?

**Comment by ryshoooo** [comment]
> Well, I think we need to support with embedded Postgres and without embedded Postgres (It could be Postgres as a service or VM. Or it could be a different type of DB (e.g. MySQL) in the future.) versions altogether.
> 
> The following settings are our current Postgres YAML structure.
> 
> ```yaml
> backendStore:
>   postgres:
>     enabled: true
>     host: "postgresql-instance1.cg034hpkmmjt.eu-central-1.rds.amazonaws.com"
>     port: 5432
>     database: "mlflow"
>     user: "mlflowuser"
>     password: "Pa33w0rd!"
> ```
> 
> Maybe we can change it to something like this for backward compatibility.
> 
> ```yaml
> backendStore:
>   embeddedPostgres:
>     enabled: true
>     ***all settings coming from [bitnami postgresql](https://artifacthub.io/packages/helm/bitnami/postgresql) chart.***
>   postgres:
>     enabled: false
>     host: ""
>     port: 5432
>     database: ""
>     user: ""
>     password: ""
> ```
> 
> What do you think?

Not sure if that's possible, AFAIK dependencies have to start with the dependency chart name as the main key in the values, i.e.
```
postgresql:
  enabled: true
  ... # values for the postgresql helm chart
backendStore:
  ....
```
I'm not sure if you can change the location. But I'd assume that if the dependency is enabled, then the backend store is not necessary to configure, it'd just deploy with the backend store being configured automatically to the postgres.

**Comment by burakince** [comment]
> Not sure if that's possible, AFAIK dependencies have to start with the dependency chart name as the main key in the values, i.e.

Yes, the dependency can start with any name including the dependency chart name. I already did it before [here](https://github.com/community-charts/helm-charts/blob/main/charts/kserve/Chart.yaml#L42).

> I'm not sure if you can change the location. But I'd assume that if the dependency is enabled, then the backend store is not necessary to configure, it'd just deploy with the backend store being configured automatically to the Postgres.

Well, we need to still keep other settings for the service or VM access configuration for production use cases. On the other hand, we can have the `postgresql` key on the top level for this dependency settings and with some complex `or` conditions, people can choose their desired backend system. Overriding subchart values from parent chart described [here](https://helm.sh/docs/chart_template_guide/subcharts_and_globals/#overriding-values-from-a-parent-chart). And probably we need to document this a bit more in the README.md file.

In the chart, we can have the following settings.

```yaml
postgresql:
  enabled: false
  # We will not set anything to here but all fields from subchart values.yaml file will be settable from here.

backendStore:
  postgres:
    enabled: false
  ...
```

And your local `values.yaml` file could be something like this.

```yaml
postgresql:
  enabled: true
  your_settings...
```

We can move the sub level postgres password and username configurations to global variables described in [here](https://helm.sh/docs/chart_template_guide/subcharts_and_globals/#global-chart-values). But it can break the backward compability.

Previously, I mentioned Bitnami's PostgreSQL helm char but I have some concerns about it.
- It uses Bitnami's docker image but we already have an official Postgres docker image. And their helm chart looks like configured too much to work with `bitnami/postgresql` docker image.
- This `bitnami/postgresql` is only built for amd64 architecture. There is no support for the ARM architecture. M1 chipset users will not be happy about this. I think we need to adapt to the ARM First Development strategy. If we can have a docker image works on minimum arm64 architecture, I think we can easily make it runnable for amd64 architecture. Official PostgreSQL docker image already supports arm64 architecture.
- It looks like too much complicated for non-production use cases.

Also, I looked at other charts but they look like not useful for us. Maybe we can create our simple PostgreSQL helm chart which will use the official docker image. And we can use it in the Mlflow helm chart. We can use global variables and adapt to them on both helm charts.

**Comment by ryshoooo** [comment]
> Yes, the dependency can start with any name including the dependency chart name. I already did it before [here](https://github.com/community-charts/helm-charts/blob/main/charts/kserve/Chart.yaml#L42).

Ah sweet, didn't know :) then sure yeah, let's do it this way :)
 
> Previously, I mentioned Bitnami's PostgreSQL helm char but I have some concerns about it.
> 
> * It uses Bitnami's docker image but we already have an official Postgres docker image. And their helm chart looks like configured too much to work with `bitnami/postgresql` docker image.
> * This `bitnami/postgresql` is only built for amd64 architecture. There is no support for the ARM architecture. M1 chipset users will not be happy about this. I think we need to adapt to the ARM First Development strategy. If we can have a docker image works on minimum arm64 architecture, I think we can easily make it runnable for amd64 architecture. Official PostgreSQL docker image already supports arm64 architecture.
> * It looks like too much complicated for non-production use cases.
> 
> Also, I looked at other charts but they look like not useful for us. Maybe we can create our simple PostgreSQL helm chart which will use the official docker image. And we can use it in the Mlflow helm chart. We can use global variables and adapt to them on both helm charts.

I would advise against :) It'll be just more maintenance burden and normally developers are used to the bitnami postgres helm chart, it's common and used in multitude of other helm charts. So more references, more documentation, and developers know how to customize it for their needs. The ARM64 missing image is not that big of a deal, I usually just overwrote the image with arm64- based image in the chart and it worked just fine. I would also advise not to do arm64-centric development as it's not that common, amd64 is still the most common architecture in use, and based on the pricing, I think it'll remain this way for a while.
